packages/astro/src/core/routing/manifest/create.ts

```diff
@@ -265,12 +273,51 @@ export function createRouteManifest(
 
 	if (fs.existsSync(pages)) {
 		walk(fileURLToPath(pages), [], []);
-	} else {
+	} else if (config?._ctx?.injectedRoutes?.length === 0) {
 		const pagesDirRootRelative = pages.href.slice(config.root.href.length);
 
 		warn(logging, 'astro', `Missing pages directory: ${pagesDirRootRelative}`);
 	}
 
+	config?._ctx?.injectedRoutes?.forEach(({pattern: name, entryPoint}) => {
```

Comment:
Moved this logic out of the `fs.existsSync(pages)` check so this will run even if `src/pages` doesn't exist